### PR TITLE
Implement type promotion in our own code

### PR DESCRIPTION
**Pull request by zasdfgbnm**
, by copy-pasting code from PyTorch, so that `DataType::Index` is correctly handled.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/type_promotion.cpp [hunk 1]
```diff
@@ -53,43 +54,73 @@ at::native::ResultTypeState updateResultTypeState(
...
+  DataType current = scalar;
+  if (scalar == DataType::Half || scalar == DataType::BFloat16) {
+    current = DataType::Float;
   }
   new_state.wrappedResult =
       promoteTypesSkipUndefined(in_state.wrappedResult, current);
   return new_state;
 }
 
+inline DataType combineCategories(DataType higher, DataType lower) {
```
Comment (zasdfgbnm) [inline csrc/type_promotion.cpp:70]: copied from pytorch

File: csrc/type_promotion.cpp [hunk 2]
```diff
@@ -53,43 +54,73 @@ at::native::ResultTypeState updateResultTypeState(
...
+  if (higher == DataType::Bool || isFloatingPointType(lower)) {
+    return promoteTypesSkipUndefined(higher, lower);
+  }
+  if (higher != DataType::Null) {
+    return higher;
+  }
+  return lower;
+}
+
+DataType resultType(const ResultTypeState& in_state) {
```
Comment (zasdfgbnm) [inline csrc/type_promotion.cpp:94]: copied from pytorch

File: csrc/type_promotion.cpp [hunk 3]
```diff
@@ -21,27 +18,31 @@ enum ValueType { Tensor, Scalar, None };
...
+DataType promoteTypesSkipUndefined(DataType a, DataType b) {
+  if (a == DataType::Null) {
     return b;
   }
-  if (b == c10::ScalarType::Undefined) {
+  if (b == DataType::Null) {
     return a;
   }
-  return c10::promoteTypes(a, b);
+  return promoteType(a, b);
```
Comment (zasdfgbnm) [inline csrc/type_promotion.cpp:38]: defined at

https://github.com/NVIDIA/Fuser/blob/fb206ef551c7ef4f1caa5d5584e297caf644fce5/csrc/type.h#L763

**Review by jjsjann123 — APPROVED** [review]
Looks like mechanical changes. Stamping.

**Pull request merged**